TF-M v1.2 integration

jainvikas8 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

PR related to upgrading TF-M v1.1 --> v1.2

The following have been changed:
* Switch from TFM-v1.1 to v1.2
* Upgrade platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/ related files
* Update post binary hook scripts for MUSCA_B1 and S1  targets
* Update partition and region_def for MUSCA_B1 and S1 targets
* Checkin new secure binaries for MUSCA_B1 and S1 targets
* Refactor code
* Remove MUSCA_A1 target support.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->
@Patater @urutva @evedon @LDong-Arm @0xc0170 
----------------------------------------------------------------------------------------------------------------

jainvikas8 committed:
Switch to using `TARGET_TFM_LATEST`

Rather than maintaining a specific `TARGET_TFM_V1_x`, its better to use
more generic name `TARGET_TFM_LATEST` to avoid confusion on the latest
TFM version supported by Mbed OS

* Rename the folder from `TARGET_TFM_V1_1` to `TARGET_TFM_LATEST`
* Update the CmakeLists.txt
* Change the name of the MUSCA targets to maintain uniformity
with TF-M v1.2
* Update target.json for PSA_V8_M to use `TFM_LATEST`

jainvikas8 committed:
Update partition files for `ARM_MUSCA_B1`

* Partition files are synced with TF-M v1.2
* To have uniformity with TF-M v1.2, rename the following:
 ** image_macros_preprocessed_ns.c to `signing_layout_ns.c`
 ** image_macros_preprocessed_s.c to `signing_layout_s.c`
* `MCUBOOT_IMAGE_NUMBER` is set to 2 by default for TF-M v1.2,
therefore it is necessary that Mbed OS compiles the right macros
for when linking and using the partition files

** Workaround **
The `region_defs.h` has an explicit definition of `BL2`, even
though it is already defined in target.json for `ARM_MUSCA_B1`.
This is because of Mbed CLI 1, as it can't seem to use the right
macro when linking the files for Mbed OS application when using
the ARMCLANG toolchain.

jainvikas8 committed:
Update `ARM_MUSCA_B1.py` post binary hook script

The script changes are required with respect to TF-M v1.2
integration for this target. The imgtool.py is been replaced with
`wrapper.py` which uses click command to run the signing algorithm.

The version `-v` and dependencies `-d` have been updated to resolve
upgrade issues from TF-M v1.1 --> v1.2

jainvikas8 committed:
Update `ARM_MUSCA_S1.py` post binary hook script

The script changes are required with respect to TF-M v1.2
integration for this target. The imgtool.py is been replaced with
`wrapper.py` which uses click command to run the signing algorithm.

The version `-v` and dependencies `-d` have been updated to resolve
upgrade issues from TF-M v1.1 --> v1.2

jainvikas8 committed:
Update target `ARM_MUSCA_S1` TF-M V1.2 binaries

* Remove the old `mcuboot.bin`
* Add `bl2.bin`
* Update TF-M secure binaries

GCC_ARM toolchain used.

jainvikas8 committed:
Remove old binary signing scripts of TF-M 1.1

jainvikas8 committed:
core: Upgrade TFM v1.2 related files

These files have been imported/copied from:
* ARMmbed/trusted-firmware-m
* ARMmbed/tf-m-tests

These are generic files, which are required for TF-M v1.2 integration
with Mbed OS for PSA_V8M and PSA_DUAL_CORE targets.

jainvikas8 committed:
Update target `ARM_MUSCA_B1` TF-M V1.2 binaries

* Remove the old `mcuboot.bin`
* Add `bl2.bin`
* Update TF-M secure binaries

GCC_ARM toolchain used.

jainvikas8 committed:
Remove: ARM_MUSCA_A1 support

ARM_MUSCA_A1 is not supported since Mbed OS 6.0
Refer: https://github.com/ARMmbed/mbed-os/pull/13165

Therefore remove files from kv_config and TF-M post binary hook script.

jainvikas8 committed:
Refactor: MUSCA targets post binary hook scripts

Raise an exception if there was an issue with handling any command
when signing the binaries for MUSCA targets.

jainvikas8 committed:
Add mbedtls_ecc_group_to_psa() to PSA in TF-M 1.2

The PSA headers imported from TF-M does not contain a declaration of
mbedtls_ecc_group_to_psa(), which is expected by pk.c from Mbed TLS.
This leads to an "undefined symbol" error when using the ARM toolchain
to compile an application for a TF-M target.

jainvikas8 committed:
Update partition files for `ARM_MUSCA_S1`

* Partition files are synced with TF-M v1.2
* To have uniformity with TF-M v1.2, rename the following:
 ** image_macros_preprocessed_ns.c to `signing_layout_ns.c`
 ** image_macros_preprocessed_s.c to `signing_layout_s.c`
* `MCUBOOT_IMAGE_NUMBER` is set to 2 by default for TF-M v1.2,
therefore it is necessary that Mbed OS compiles the right macros
for when linking and using the partition files.

evedon commented on platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_V1_2/include/psa/crypto_extra.h:
```diff
@@ -57,6 +57,74 @@ extern "C" {
 
 /**@}*/
 
+#if defined(MBEDTLS_ECP_C)
```
Could we put all Mbed OS platform specific files in a dedicated `platform` folder as we do for mbedtls in https://github.com/ARMmbed/mbed-os/tree/master/connectivity/mbedtls/platform?

LDong-Arm replied:
This is not platform specific - `MBEDTLS_ECP_C` is an Mbed TLS macro. This header is a standard part of Mbed TLS, but the copy maintained in TF-M's codebase (which we copy here) is slightly out of date, thus we made this change to align with Mbed TLS 2.24.0.

As for file paths, `connectivity/mbedtls/` is for non-PSA functionalities, and `platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_V1_2/include/psa/` is for PSA functionalities.

LDong-Arm reviewed (commented):
Looks good to me. Will approve once the license headers of the imported scripted are resolved and it's rebased on #14160.

jainvikas8 committed:
Add missing license header and SPDX identifier

This files are imported and part of:
https://github.com/mcu-tools/mcuboot/commit/a8e12dae381080e898cea0c6f7408009b0163f9f

Issue has been raised https://github.com/mcu-tools/mcuboot/issues/930

As these scripts are going to be part of release its important to have
licensing information.

evedon commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_B1/partition/region_defs.h:
```diff
@@ -21,6 +21,10 @@
 #include "flash_layout.h"
 
+#ifndef BL2
+#define BL2
```
Have you captured somewhere that this is a workaround for Mbed CLI 1 that we will nned to remove when migrating to Mbed CLI 2?

jainvikas8 replied:
Issue logged

LDong-Arm reviewed (approved):
The rebase and license headers look good since my previous review.

urutva commented on .gitignore:
```diff
@@ -99,7 +99,8 @@ test_suite.json
 # Directory used to clone and build TF-M
-features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_IGNORE/
+platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_V1_0/TARGET_IGNORE/
+platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_V1_2/TARGET_IGNORE/
```
It's better to avoid TF-M version in the path name as this will lead significant maintenance overhead when updating to future version of TF-M. I'd recommend using `TARGET_TFM_LATEST` or something alike to avoid moving files around and the yaml file (https://github.com/ARMmbed/mbed-os-tf-m-regression-tests/blob/master/tfm_ns_import.yaml).

jainvikas8 replied:
Yes, makes sense to switch to TARGET_TFM_LATEST

urutva commented on tools/targets/ARM_MUSCA_B1.py:
```diff
@@ -109,3 +140,12 @@ def find_bl2_size(configFile):
+        cwd=directory,
+    )
+
+    POPEN_INSTANCE.communicate()
```
`run_cmd` assumes that signing will always succeed. It should check at least `returncode` to see if process succeeded. If failed then tap into `stderr` and print it on the console, if not then failure information will be lost.

jainvikas8 replied:
A good point will look into.

urutva commented on targets/targets.json:
```diff
@@ -4686,7 +4686,8 @@
             "LPTICKER_DELAY_TICKS=3",
             "MBED_MPU_CUSTOM",
-            "BL2"
+            "BL2",
```
Do we need `#define BL2` for Musca S1 in `region_defs.h` similar to https://github.com/ARMmbed/mbed-os/pull/14187/commits/9c31de2eafe738454ff4a8ee29cac374374caaf3#diff-f120407ee71d031cc2ad72a6de1da133946f8681f166f180c5a7b7c8bace401fR25? Or Musca S1 is not affected?

jainvikas8 replied:
MUSCA S! already has it... this was done as part of https://github.com/ARMmbed/mbed-os/pull/13196/commits/37f26692b1db388a272c9037b28cad27734fda90

urutva commented on tools/targets/ARM_MUSCA_S1.py:
```diff
@@ -109,3 +140,12 @@ def find_bl2_size(configFile):
+        cwd=directory,
+    )
+
+    POPEN_INSTANCE.communicate()
```
`run_cmd` assumes that signing will always succeed. It should check at least `returncode` to see if process succeeded. If failed then tap into `stderr` and print it on the console, if not then failure information will be lost.

LDong-Arm reviewed (approved):
LGTM

LDong-Arm reviewed (approved):
LGTM. GitHub shows the final result of the force-push is unchanged, so it should be good.
